parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java

```diff
@@ -256,6 +276,28 @@ public Builder withDictionaryPageSize(int dictionaryPageSize) {
       return this;
     }
 
+    /**
+     * Set to enable Bloom filter.
+     *
+     * @param enableBloomFilter a boolean to indicate whether to enable Bloom filter.
+     * @return this builder for method chaining.
+     */
+    public Builder withBloomFilterEnabled(boolean enableBloomFilter) {
+      this.enableBloomFilter = enableBloomFilter;
+      return this;
+    }
+
+    /**
+     * Set Bloom filter info for columns.
+     *
+     * @param columnExpectedNDVs the columns expected number of distinct values in a row group
+     * @return this builder for method chaining
+     */
+    public Builder withBloomFilterInfo(HashMap<String, Long> columnExpectedNDVs) {
```

Comment:
So, the user is able to specify the NDV for the columns which he would like to write bloom filters for. If this map is `null` or empty then no bloom filters should be written otherwise bloom filters are written only for the columns which are NDV is specified for. 
Why do we need the additional flag of enabling bloom filters?
Is there any other way to build a bloom filter without a specified NDV? Usually, parquet-mr does not ask for such values about the data characteristics but measures them. I don't know how would the clients present this data and therefore if they willing to use this API.

Reply:
You are right,  if user doesn't specific any NDV, no bloom filter will be build according to current logic. The additional flag is a redundant flag.

Reply:
I'm not all that familiar with the internals of bloom filters but was hoping this could work similar to dictionary fallback where Parquet automatically uses a dictionary provided the size of distinct values does not grow beyond the specified max dictionary size e.g. dictionaries would be preferred but would fallback to bloom filters for higher cardinality fields. Then again, it doesn't make sense to probably try to do this for all columns automatically (so opt-in) although it'd be nice if there were some heuristics that could decide that automatically. It also looks like that may not be possible though because the # of distinct values needs to be known ahead of time to size this correctly? If that number is incorrect it looks like it increases the likelihood of false positives and if it's too large it takes up more space than needed?

I looked briefly into how this works for ORC where the bloom filter columns are opt-in specified via column name (without providing unique counts). But, the reason they're able to omit the # of distinct values is because the bloom filters are created at the `stride` level where ORC controls the stride size by number of rows so they always just size it to the stride row size (default of 10k). So, they assume worst case scenario that all values are unique within the stride (so the filter ends up being a larger size than it needs to be maybe but without the false positive probability increasing).

I'm not sure if it will be problematic here to require clients to specify the number of distinct values upfront but I can't really think of a better solution at the moment (short of something painful/inefficient like Parquet doing two passes of the data).

Reply:
Thanks @ryanrupp , I know the FPP may not close to expected value since user may not estimate NDV correctly before row group size is determined.  I used to think cache hashed value and build bloom filter when writing column chunk, that way can keep FPP as we expected and can take advantage some vector computation while it may lead to OOM in case of very large NDV.  Another option should be build bloom filter on page level so the behavior can be same as ORC side but that may make it more complicated.

Specify NDV looks like a bit difficult to user, so I 'm thinking to add another property named max size of bloom filter of column, so that users can specify how much space they would like to use as bloom filter index for a column. It looks simple and direct to users. What do you think?

Reply:
I agree with @ryanrupp, the current behavior that the user have to specify the NDVs is somewhat against the usual way how _parquet-mr_ works. If it is hard to guess such values the clients may not use this feature and the whole development is useless. 
Also, I think such discussions (about how the library would guess the parameters of a bloom filter or which columns should we build one) would more fit in a design doc.
However, the current development is on the feature branch so, for the sake of incremental development, if it is easier for you to start the implementation and/or execute benchmarks, I can accept the current way for now.

Reply:
Thanks @gszadovszky @ryanrupp , I opened a [discussion](https://lists.apache.org/thread.html/f783f4a6e4e98f839285e78d0fa27db05628348e28a6608791fba740@%3Cdev.parquet.apache.org%3E) on dev mail list, welcome to put your opinion.  I will update design doc and implementation if we have some agreement.